### PR TITLE
spirv-fuzz: Add illustrative tests for new issues

### PR DESCRIPTION
**Pull request by afd**
Adds (currently disabled) tests related to #4345 and #4346.

### CODE REVIEW

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp [hunk 1]
```diff
@@ -1796,6 +1796,238 @@ TEST(TransformationReplaceIdWithSynonymTest, IncompatibleTypes) {
...
+  transformation_context.GetFactManager()->AddFactDataSynonym(
+      MakeDataDescriptor(101, {}), MakeDataDescriptor(20, {}));
+
+  {
+    const auto& scope_operand = MakeIdUseDescriptorFromUse(
+        context.get(), context->get_def_use_mgr()->GetDef(21), 1);
+    TransformationReplaceIdWithSynonym replace_scope(scope_operand, 100);
+    ASSERT_TRUE(
+        replace_scope.IsApplicable(context.get(), transformation_context));
+    replace_scope.Apply(context.get(), &transformation_context);
```
Comment (Vasniktel) [inline test/fuzz/transformation_replace_id_with_synonym_test.cpp:1973]: Probably not necessary for this transformation, but I recall we've been using `ApplyAndCheckFreshIds` everywhere.

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp [hunk 2]
```diff
@@ -1796,6 +1796,240 @@ TEST(TransformationReplaceIdWithSynonymTest, IncompatibleTypes) {
...
+        context.get(), context->get_def_use_mgr()->GetDef(21), 2);
+    TransformationReplaceIdWithSynonym replace_semantics(semantics_operand,
+                                                         101);
+    ASSERT_TRUE(
+        replace_semantics.IsApplicable(context.get(), transformation_context));
+    ApplyAndCheckFreshIds(replace_semantics, context.get(),
+                          &transformation_context);
+  }
+
+  ASSERT_TRUE(fuzzerutil::IsValidAndWellFormed(context.get(), validator_options,
```
Comment (Vasniktel) [inline test/fuzz/transformation_replace_id_with_synonym_test.cpp:1988]: Probably better to move this into those 2 scopes.

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp (hunk 1 above)
Comment (afd) [reply test/fuzz/transformation_replace_id_with_synonym_test.cpp]: You're right - fixed.

**Review by Vasniktel — APPROVED** [review]

File: test/fuzz/transformation_replace_id_with_synonym_test.cpp (hunk 2 above)
Comment (afd) [reply test/fuzz/transformation_replace_id_with_synonym_test.cpp]: Done.

Comment (afd) [reply test/fuzz/transformation_replace_id_with_synonym_test.cpp]: Done.

**Pull request merged**